Cooler functionality

mozesjacobs opened the pull request:
- Created cooler.py
- modified all_data.py, pastis-poisson, and pastis-algorithms.py
- _filter_.c got modified, and so did fastio_.c

gesinecauer commented:
Hey @NelleV! We had a couple questions for you:

(1) What's the best way to specify which type of diploid contact counts files are being inputted?
When loading hiclib data, the code distinguishes unambiguous from partially ambiguous from ambiguous by the shape of the matrix (relative to the chromosome lengths). This allows a list of counts to be inputted via the `--counts` argument.
However, cooler stores diploid data differently - there would be one file for 'maternal-maternal' contacts, another for 'maternal-paternal', 'maternal-unknown', etc etc. Naturally, there is no way for the code to distinguish between these based on shape. We considered a few different ways of managing this:
1. We could have separate arguments for each type of counts matrix (ie `--counts_hiclib`, `--counts_cooler_maternal_maternal`, `--counts_cooler_maternal_paternal`, etc.)
2. We could have a single argument for the counts, as we did before, and have the type of counts indicated by the filename (`counts.matrix`, `counts.maternal-maternal.cool`, `counts.maternal-paternal.cool`, etc.)
3. We could create a standalone script that converts from the cooler to hiclib formats, and only allow hiclib-formatted files to be inputted to pastis
4. We could have two separate scripts to run pastis, depending on whether hiclib or cooler files are inputted
What do you think would be best?

(2) What is the best way to allow for importing of the cooler package when necessary, without having cooler as a dependency?

Thanks! :)

gesinecauer commented:
(By the way, the tests are currently failing because Travis isn't installing cooler - we will fix this)

NelleV commented:
Hi @mozesjacobs 

Thanks for the PR. I'll do the review a bit later: for now, we need to make sure that travis passes. Unfortunately, there's two problems for this.

1. The numpy version in the travis configuration has been bumped down to 1.14. We need it back to 1.16, which is the minimun numpy version we are supporting.
2. The compiled files have been regenerated. Unfortunately, I think you have a pretty old version of numpy and it's now causing issues with recent version of numpy. These also don't need to be regenerated. A fix could be to revert those to their original state: this would make travis pass.

Let me know if you need any help with this.
N

mozesjacobs commented:
Hi @NelleV 
Thank you for the help! I changed the numpy version in travis, and then I reverted the versions of _filter_.c and fastio_.c to the commit before I started working (d24d8deb99f54dbde4694c5f916fc4ee06126780 with message "MAINT version 0.4 has been released"). I'm not sure what's wrong though, since Travis is still failing. Is there a step I have missed?

gesinecauer commented on .travis.yml:
```diff
@@ -12,13 +12,13 @@ matrix:
     - name: "Python 3.6"
       python: 3.6
-      env: TEST=TRUE NUMPY_VERSION=1.16 SCIPY_VERSION=0.19  PANDAS_VERSION=0.25 SKLEARN_VERSION=0.18 AUTOGRAD_VERSION=1.2 COVERAGE=true
+      env: TEST=TRUE NUMPY_VERSION=1.14 SCIPY_VERSION=0.19  PANDAS_VERSION=0.25 SKLEARN_VERSION=0.18 AUTOGRAD_VERSION=1.2 COOLER_VERSION=0.8.10 COVERAGE=true
```
Oh shoot! I hadn't updated gesinecauer/pastis with hiclib/pastis in a while, so the code I suggested Mozes replace was outdated. Sorry about that!

gesinecauer commented:
@mozesjacobs, it looks like the tests are now failing because the python API has changed, which is not too surprising.

I'll send a message to Bill about whether we should have the cooler-loading functionality be integrated or a standalone script, and then we can move forward on this. If we integrate cooler-loading functionality, I'll probably have to update my unit tests

mozesjacobs commented:
@gesinecauer okay, sounds good!

NelleV commented:
Hi @mozesjacobs 
Thanks for fixing the travis-CI build! I think we're almost there.
Here's a list of small elements that we need to fix before merging the PR:
1. the script should be moved to `pastis/script` as all the other scripts included in pastis. This are not typical submodules so we keep them separate.
2. in order for the script to be installed properly, you need to edit `setup.py` in the root folder. Search for "scripts" in this file. Reading the lines of code should give you an idea of what needs to be edited for proper installation.
3. I think we should add a very small test script that checks the conversion. In order to do this, I suggest you have a look at how I test the `iced` script in the iced package. 
     a. In the example folder, I wrote a small bash script with a call to the script, using the same options as what we use in HiC-pro (see https://github.com/hiclib/iced/tree/master/examples/HiC-pro). In iced, the issue came from writing the normalized contact count matrix, so I also wrote a small python script that loads the script back in. We could do something similar here.
     b. Once this small bash script works and you've made sure it tests the script sufficiently to pick up any issues, you have to edit the scripts ran by travis-CI in order for the this script to be ran. This is in the `build_tools/travis` folder, and you should check out the test_script.sh. You can look at how it's done on `iced` to see what I modified to execute the tests on the scripts: https://github.com/hiclib/iced/blob/master/build_tools/travis/test_script.sh

Let me know if any of the steps are unclear,
Nelle
